srtcore/atomic.h

```diff
@@ -61,16 +61,56 @@
       line)[(2 * static_cast<int>(!!(condition))) - 1] _impl_UNUSED
 #endif
 
-#if defined(__GNUC__) || defined(__clang__) || defined(__xlc__)
-#define ATOMIC_USE_GCC_INTRINSICS
+#if defined(ATOMIC_USE_POSIX_MUTEX) && (ATOMIC_USE_POSIX_MUTEX == 1)
+   // NOTE: Defined at the top level.
+#elif defined(__APPLE__) && (__cplusplus >= 201103L)
+   // NOTE: Does support c++11 std::atomic, but the compiler may or
+   //    may not support GCC atomic intrinsics. So go ahead and use the
+   //    std::atomic implementation.
+   #define ATOMIC_USE_CPP11_ATOMIC
+#elif (defined(__clang__) && defined(__clang_major__) && (__clang_major__ > 5)) \
+   || defined(__xlc__)
+   // NOTE: Clang <6 does not support GCC __atomic_* intrinsics. I am unsure
+   //    about Clang6. Since Clang sets __GNUC__ and __GNUC_MINOR__ of this era
+   //    to <4.5, older Clang will catch the setting below to use the
+   //    POSIX Mutex Implementation.
+   #define ATOMIC_USE_GCC_INTRINSICS
+#elif defined(__GNUC__) \
+   && ( (__GNUC__ > 4) || ((__GNUC__ == 4) && (__GNUC_MINOR__ >= 7)) )
+   // NOTE: The __atomic_* family of intrisics were introduced in GCC-4.7.0.
+   // NOTE: This follows #if defined(__clang__), because most if, not all,
+   //    versions of Clang define __GNUC__ and __GNUC_MINOR__ but often define
+   //    them to 4.4 or an even earlier version. Most of the newish versions
+   //    of Clang also support GCC Atomic Intrisics even if they set GCC version
+   //    macros to <4.7.
+   #define ATOMIC_USE_GCC_INTRINSICS
+#elif defined(__GNUC__) && !defined(ATOMIC_USE_POSIX_MUTEX)
+   // NOTE: GCC compiler built-ins for atomic operations are pure
+   //    compiler extensions prior to GCC-4.7 and were grouped into the
+   //    the __sync_* family of functions. GCC-4.7, both the c++11 and C11
+   //    standards had been finalized, and GCC updated their built-ins to
+   //    better reflect the new memory model and the new functions grouped
+   //    into the __atomic_* family. Also the memory models were defined
+   //    differently, than in pre 4.7.
+   // TODO: PORT to the pre GCC-4.7 __sync_* intrinsics. In the meantime use
+   //    the POSIX Mutex Implementation.
+   #define ATOMIC_USE_POSIX_MUTEX 1
 #elif defined(_MSC_VER)
-#define ATOMIC_USE_MSVC_INTRINSICS
-#include "atomic_msvc.h"
+   #define ATOMIC_USE_MSVC_INTRINSICS
+   #include "atomic_msvc.h"
 #elif __cplusplus >= 201103L
-#define ATOMIC_USE_CPP11_ATOMIC
-#include <atomic>
+   #define ATOMIC_USE_CPP11_ATOMIC
 #else
-#error Unsupported compiler / system.
+   #error Unsupported compiler / system.
+#endif
+// Include any necessary headers for the selected Atomic Implementation.
+#if defined(ATOMIC_USE_POSIX_MUTEX) && (ATOMIC_USE_POSIX_MUTEX == 1)
+   #include <assert.h>
+   #include <stdio.h>
```

Comment:
```suggestion
   #include <cassert>
   #include <cstdio>
```

Reply:
No longer needed with the latest commit.